crates/matrix-sdk-indexeddb/src/stream.rs

```diff
@@ -0,0 +1,624 @@
+        // The `latest_transaction`.
+        #[pin]
+        latest_transaction: IdbTransaction<'a>,
+
+        // The self-reference to `Self::latest_transaction`.
+        latest_transaction_ptr: NonNull<IdbTransaction<'a>>,
+
+        // The `latest_object_store`.
+        #[pin]
+        latest_object_store: Option<IdbObjectStore<'a>>,
+
+        // The self-reference to `Self::latest_object_store`.
+        latest_object_store_ptr: NonNull<Option<IdbObjectStore<'a>>>,
+
+        // The inner stream, aka [`StreamByCursor`].
+        #[pin]
+        inner_stream: Option<StreamByCursor<'a>>,
+
+        // When asking for a new cursor, the [`IdbObjectStore::open_cursor_with_range`]
+        // method has to be called. It's an async method. Thus, in the `Stream`
+        // implementation of `Self`, this future must be stored to be polled manually.
+        cursor_future: Option<Pin<Box<IdbCursorWithValueFuture<'a, IdbObjectStore<'a>>>>>,
+
+        // The entire struct must be unmovable. Let's use a `PhantomPinned` so that it cannot implement
+        // `Unpin`.
+        _pin: PhantomPinned,
+    }
+}
+
+impl<'a, F> StreamByRenewedCursor<'a, F>
+where
+    F: FnMut() -> Result<IdbTransaction<'a>, DomException>,
+{
+    /// Build a new `StreamByRenewdCursor`.
+    ///
+    /// It takes a `transaction_builder`, an `object_store_name`, and a
+    /// `renew_every`. See the documentation of [`Self`] to learn more.
+    pub async fn new(
+        mut transaction_builder: F,
+        object_store_name: String,
+        renew_every: NonZeroUsize,
+    ) -> Result<Pin<Box<Self>>, DomException> {
+        let transaction = transaction_builder()?;
+        let latest_key = JsValue::from_str("");
+        let after_latest_key = IdbKeyRange::lower_bound_with_open(&latest_key, true)?;
+
+        let this = Self {
+            transaction_builder,
+            object_store_name,
+            renew_every: renew_every.into(),
+            renew_in: renew_every.into(),
+            latest_key,
+            latest_transaction: transaction,
+            latest_transaction_ptr: NonNull::dangling(),
+            latest_object_store: None,
+            latest_object_store_ptr: NonNull::dangling(),
+            inner_stream: None,
+            cursor_future: None,
+            _pin: PhantomPinned,
+        };
+        let mut this = Box::pin(this);
+
+        unsafe {
+            let this = Pin::get_unchecked_mut(Pin::as_mut(&mut this));
+
+            this.latest_transaction_ptr = NonNull::from(&this.latest_transaction);
+            this.latest_object_store = Some(
+                this.latest_transaction_ptr
+                    .as_ref()
+                    .object_store(this.object_store_name.as_str())?,
+            );
+            this.latest_object_store_ptr = NonNull::from(&this.latest_object_store);
+
+            let cursor = this
+                .latest_object_store_ptr
+                .as_ref()
+                .as_ref()
+                .unwrap()
```

Comment:
Worth a comment that `unwrap` is safe because we just created this thing and made a pointer to it?